### PR TITLE
Maker reliability updates

### PR DESCRIPTION
**Pull request by dbelv**
- Improve the logging output to be readable from the terminal. If `RUST_LOG` is not set use a global level of `warn` and a maker level of `info`.
- Improve reliability around error cases.
- Add in `rustfmt` file to set the max lines to 120 over the default of 80.
- Improve logging around reconnection events with the Valorem server.
- Sort out the `Cargo.toml` file correcting which dependencies are dev and which are not.
- Remove todo around DRY and use a wrapping enum for creation of the `ethers-rs` providers.
- Panics still cut through the 3rd party library, given additional changes remove the 3rd party library and instead opt for a simple `loop` and `sleep` structure.
- No need to call out to the chain to fetch the block timestamp, seaport order timestamps don't need to be that accurate.

### CODE REVIEW

File: examples/rust/examples/taker/taker.rs [hunk 1]
```diff
@@ -416,9 +385,7 @@ async fn setup_valorem_connection<P: JsonRpcClient + 'static>(
     match response {
         Ok(_) => (),
         Err(error) => {
-            eprintln!(
-                "Error: Unable to check authentication with Valorem. Reported error:\n{error:?}"
-            );
+            eprintln!("Error: Unable to check authentication with Valorem. Reported error:\n{error:?}");
```
Comment (0xAlcibiades) [inline examples/rust/examples/taker/taker.rs]: Favor tracing crate.

File: examples/rust/examples/taker/taker.rs [hunk 2]
```diff
@@ -499,15 +462,10 @@ async fn setup_option<P: JsonRpcClient + 'static>(
...
-            bindings::valorem_clear::SettlementEngineEvents::decode_log(&RawLog { topics, data })
-                .unwrap();
+        let event = bindings::valorem_clear::SettlementEngineEvents::decode_log(&RawLog { topics, data }).unwrap();
 
         if let bindings::valorem_clear::SettlementEngineEvents::NewOptionTypeFilter(event) = event {
-            println!(
-                "Option Id successfully created. Option Id {:?}",
-                event.option_id
-            );
+            println!("Option Id successfully created. Option Id {:?}", event.option_id);
```
Comment (0xAlcibiades) [inline examples/rust/examples/taker/taker.rs]: tracing

File: examples/rust/rustfmt.toml [hunk 3]
```diff
@@ -0,0 +1 @@
+max_width = 120
```
Comment (0xAlcibiades) [inline examples/rust/rustfmt.toml]: Eh, I mean, if we are going to do this, let's do this on another PR so the delta here is more readable.

File: examples/rust/examples/maker/tracing_utils.rs
Comment (0xAlcibiades) [inline examples/rust/examples/maker/tracing_utils.rs:1]: Let's keep this, tracing is desireable, it let's us better know what's going on.

File: examples/rust/Cargo.toml [hunk 4]
```diff
@@ -44,19 +18,28 @@ version = "1.0.160"
...
-features = ["serde"]
+[dependencies.tonic]
+version = "0.9.1"
+features = ["tls"]
 optional = false
 
-[dependencies.tokio]
-version = "1.27"
-features = ["macros", "rt-multi-thread"]
+[dev-dependencies]
```
Comment (0xAlcibiades) [inline examples/rust/Cargo.toml:26]: Should these be dev or example dependencies?

File: examples/rust/examples/maker/maker.rs [hunk 5]
```diff
@@ -27,7 +25,12 @@ use valorem_trade_interfaces::{
 };
 
 mod settings;
-mod tracing_utils;
```
Comment (0xAlcibiades) [inline examples/rust/examples/maker/maker.rs:30]: Let's use tracing.

File: examples/rust/examples/maker/maker.rs [hunk 6]
```diff
@@ -38,100 +41,117 @@ const TOS_ACCEPTANCE: &str = "I accept the Valorem Terms of Service at https://a
...
+
+    // Initialise a coloured timed logger
+    pretty_env_logger::init_timed();
 
-    // TODO(DRY)
     let args: Vec<String> = env::args().skip(1).collect();
     if args.len() != 1 {
-        eprintln!("Unexpected command line arguments. Received {:?}", args);
-        eprintln!("Usage: maker <settings_file>");
+        error!("Unexpected command line arguments. Received {:?}", args);
```
Comment (0xAlcibiades) [inline examples/rust/examples/maker/maker.rs:54]: tracing.

File: examples/rust/examples/maker/maker.rs [hunk 7]
```diff
@@ -1,21 +1,19 @@
...
 use ethers::abi::{AbiEncode, RawLog};
 use ethers::prelude::{
     rand::{thread_rng, Rng},
-    Address, EthLogDecode, Http, JsonRpcClient, LocalWallet, Middleware, Provider, Signer,
+    Address, EthLogDecode, Http, Ipc, JsonRpcClient, LocalWallet, Middleware, Provider, Signer,
     SignerMiddleware, Ws, U256,
 };
 use ethers::utils::keccak256;
 use http::Uri;
+use log::{error, info, warn};
```
Comment (0xAlcibiades) [inline examples/rust/examples/maker/maker.rs:10]: Let's use tracing.

**Review by 0xAlcibiades — CHANGES_REQUESTED** [review]
Look good, but let's use tracing, unless some good rationale why logging is better, and let's bring in the line width change on another PR.

**Review by 0xAlcibiades — APPROVED** [review]

File: examples/rust/examples/maker/maker.rs (hunk 7 above)
Comment (megsdevs) [reply examples/rust/examples/maker/maker.rs:10]: agree for consistency, unless there is some reason this comment has been marked resolved @0xAlcibiades?

File: examples/rust/examples/maker/maker.rs [hunk 8]
```diff
@@ -38,100 +41,117 @@ const TOS_ACCEPTANCE: &str = "I accept the Valorem Terms of Service at https://a
...
+        // We never expect the run function above to end, however it can primarily due to the connection to
+        // Valorem being closed. Sleep for a set period of time in case there are other issues at play
+        // (i.e. Valorem down) before retrying to connect.
+        warn!("Reconnection Event: Returned from the main run function, reconnecting to Valorem");
+        sleep(Duration::from_secs(10)).await;
+    }
+}
+
+// Helper function to connect to a node provider.
+async fn connect_to_node_provider(node_endpoint: String) -> EthersProvider {
```
Comment (megsdevs) [inline examples/rust/examples/maker/maker.rs:85]: nice

File: examples/rust/examples/maker/maker.rs [hunk 9]
```diff
@@ -38,100 +41,117 @@ const TOS_ACCEPTANCE: &str = "I accept the Valorem Terms of Service at https://a
...
-
-    Ok(())
 }
 
 // Main execution function. This is not expected to return.
 async fn run<P: JsonRpcClient + 'static>(
     provider: Arc<Provider<P>>,
     settings: Settings,
-) -> Result<(), anyhow::Error> {
+) -> Option<()> {
```
Comment (megsdevs) [inline examples/rust/examples/maker/maker.rs:134]: losing error info here which would be useful for debugging. how do we no it's not continually hitting an error and reconnecting?

File: examples/rust/examples/maker/maker.rs [hunk 10]
```diff
@@ -38,100 +41,117 @@ const TOS_ACCEPTANCE: &str = "I accept the Valorem Terms of Service at https://a
...
+            .tls_config(settings.tls_config.clone())
+            .ok()?
             .http2_keep_alive_interval(Duration::new(75, 0))
             .keep_alive_timeout(Duration::new(10, 0))
             .timeout(Duration::from_secs(10))
             .connect_timeout(Duration::from_secs(10))
             .connect()
-            .await?,
+            .await
+            .ok()?,
```
Comment (megsdevs) [inline examples/rust/examples/maker/maker.rs]: be nice if all errors which cause restart are logged rather than silent

File: examples/rust/examples/maker/maker.rs [hunk 11]
```diff
@@ -143,12 +163,14 @@ async fn run<P: JsonRpcClient + 'static>(
...
         SignerMiddleware::new_with_provider_chain(Arc::clone(&provider), settings.wallet.clone())
             .await
-            .unwrap();
+            .ok()?;
 
     // Seaport 1.5 contract address
+    // Note: We allow the unchecked unwrap here, since this address will always parse correctly.
     let seaport_contract_address = "0x00000000000000ADc04C56Bf30aC9d3c0aAF14dC"
         .parse::<Address>()
         .unwrap();
```
Comment (megsdevs) [inline examples/rust/examples/maker/maker.rs:178]:
```suggestion
        .ok?;
```

File: examples/rust/examples/maker/maker.rs (hunk 9 above)
Comment (megsdevs) [reply examples/rust/examples/maker/maker.rs:134]: i would probably keep as result, propagate the errors with `?`s, then log the error in the reconnect loop

File: examples/rust/examples/maker/maker.rs [hunk 12]
```diff
@@ -168,9 +189,11 @@ async fn run<P: JsonRpcClient + 'static>(
...
             .await
-            .unwrap()
+            .ok()?
             .into_inner();
 
+        info!("Ready for RFQs from Takers");
+
         // Now we have received the RFQ request stream - loop until it ends.
         while let Ok(Some(quote)) = quote_stream.message().await {
             // Check the chain-id is valid
```
Comment (megsdevs) [inline examples/rust/examples/maker/maker.rs]: the server already does this now, but extra safe to keep here

Comment (megsdevs) [reply examples/rust/examples/maker/maker.rs]: same for ulid check done in validate_rfq, but makers should be doing these checks as well, so keep here

**Review by megsdevs — CHANGES_REQUESTED** [review]
Overall very clean, some changes requested:
- main run function to return result instead of option so all the errors within run can be propagated. At the moment there are many silent errors that could occur. Then log the error in the restart loop.
- there are still some unwraps to deal with (do a cmd F to be thorough)

**Comment by 0xAlcibiades** [comment]
@megsdevs I resolved the tracing comments as I think a tracing instrumented version for an example maker is out of scope, after further thought on it.

File: examples/rust/Cargo.toml (hunk 4 above)
Comment (dbelv) [reply examples/rust/Cargo.toml:26]: Dev-dependencies are not used when compiling a package for building, but are used for compiling tests, examples, and benchmarks.

These changes were made since the dependencies were incorrect in the exist toml, only those now contained under dependencies are required.

File: examples/rust/examples/maker/maker.rs (hunk 7 above)
Comment (dbelv) [reply examples/rust/examples/maker/maker.rs:10]: So I avoided tracing since the output is not as easy to review when running as an example (so I wouldn't expect anything to be required to ingest and decode logs). So this logging change moved the trace from:
```
{"v":0,"name":"trade-api-maker","msg":"Re/starting maker","level":40,"hostname":"Moria","pid":19125,"time":"2023-10-22T23:10:45.200577289Z","target":"maker","line":57,"file":"examples/maker/maker.rs"}
{"v":0,"name":"trade-api-maker","msg":"Maker has authenticated with Valorem","level":30,"hostname":"Moria","pid":19125,"time":"2023-10-22T23:10:48.394038536Z","target":"maker","line":735,"file":"examples/maker/maker.rs"}
{"v":0,"name":"trade-api-maker","msg":"Ready for RFQs from Takers","level":30,"hostname":"Moria","pid":19125,"time":"2023-10-22T23:10:49.222303679Z","target":"maker","line":162,"file":"examples/maker/maker.rs"}
```

to 

```
 2023-10-22T23:12:38.741Z INFO  maker > Maker has authenticated with Valorem
 2023-10-22T23:12:39.978Z INFO  maker > Ready for RFQs from Takers
```

This is far more better to read while running the example.

Not sure what you mean by consistency here, the Taker example uses println and the library doesn't have any logging.

File: examples/rust/examples/maker/maker.rs (hunk 11 above)
Comment (dbelv) [reply examples/rust/examples/maker/maker.rs:178]: This is never expected to fail so we should never silently ignore this, we want this to panic since its a hardcoded address (i.e. we can only fix it up updating code).

File: examples/rust/examples/maker/maker.rs (hunk 12 above)
Comment (dbelv) [reply examples/rust/examples/maker/maker.rs]: Never assume any input to an application you are writing has been checked/is correct. You should always sanitize any input data you program receives before dealing with it.

**Comment by dbelv** [comment]
>     * there are still some unwraps to deal with (do a cmd F to be thorough)

There are only a couple of places where by unwraps are kept. 

- Line 178: We always want this line to panic if it fails since its a hardcoded address
- Line 221, 283, 331, 365: These are checked unwraps
- ~Line 410: We want this to panic, since if the provider cannot fetch the chain id it signals the node is bad. While we could have automatic restarts on node issues, this was done to alert the user there are issues as we don't really expect the example to be running for days.~
- Line 504, 509, 510, 512: These are unchecked since we built the structure we are unwrapping from
- Line 552: This needs to panic, since that means something is wrong with Ether-rs as its no longer able to sign data from a wallet. This is basically an internal error from Ethers.
- Line 579, 582: These are checked unwraps
- Line 722, 725: These are hardcoded values, we want these to panic if they fail.
- ~Line 708: Same reasoning as Line 410.~
- Line 741: This is the same reasoning as Line 517
- The unchecked unwraps within the `approve_token` function are there deliberately as you immediately want to fail and fail hard if there are issues with approving tokens as you always want the user to immediately review.
- Line 918: If this fails the system you are running on is likely about to fail.

Edit: On further thought we should be forcing a reconnect with when we cannot get the chain_id

File: examples/rust/examples/maker/maker.rs (hunk 9 above)
Comment (dbelv) [reply examples/rust/examples/maker/maker.rs:134]: > how do we no it's not continually hitting an error and reconnecting?

The main loop will be logging if it was forced to reconnect.

Added additional context around reconnection events.

**Comment by dbelv** [comment]
>     * main run function to return result instead of option so all the errors within run can be propagated. At the moment there are many silent errors that could occur. Then log the error in the restart loop.

I don't think this is the correct way to handle errors as a general practice (having everything propagated up individually). Ideally you always handle errors as close to the event where it could occur and then either convert the errors to a common overall type/enum (with any additional context) so its quite clear how the error is going to be handled in the top level/caller or immediately panic/crash.

I have added logging around all reconnection events.

File: examples/rust/examples/maker/maker.rs (hunk 10 above)
Comment (dbelv) [reply examples/rust/examples/maker/maker.rs]: Done

File: examples/rust/examples/maker/maker.rs (hunk 7 above)
Comment (megsdevs) [reply examples/rust/examples/maker/maker.rs:10]: in that case leave as is, i meant by trade api, but you are correct for the examples
